### PR TITLE
CSHARP-2651: Update change stream docs example for resume token access

### PR DESCRIPTION
**Pull request by MikalaiMazurenka**
Evergreen: https://evergreen.mongodb.com/version/5da4e50fe3c331507661fe80

### CODE REVIEW

File: tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs
Comment (rstam) [inline tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs]: The fact that this changed from 1 to 2 means there is a bug in the sample code.

This example code is terrible... but that's not your fault of course. It was terrible when it was first written. No one should be consuming a change stream this way.

But I think the changes you have made have introduced a bug.

By mixing the use of `IEnumerator` and `IAsyncCursor` you have ended up with the wrong resume token.

I can't really tell from the JIRA ticket what you are supposed to change. How did you determine what needed to be done?

**Review by rstam — CHANGES_REQUESTED** [review]

File: tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs
Comment (vincentkam) [inline tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs:116]: I was curious where these examples were actually being used as we don't appear to be referencing this example in our own ChangeStream documentation (in our reference documentation). Tracing the history of this ticket, it appears that these examples are meant to be consumed by the docs team (see https://jira.mongodb.org/browse/DRIVERS-414) and the code for the original example 3 eventually ended up here:  https://docs.mongodb.com/manual/changeStreams/#resume-a-change-stream.

Comment (vincentkam) [reply tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs:116]: @jyemin After Java finished their ticket, was a separate docs ticket needed to update the examples at docs.mongodb.com?

File: tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs
Comment (vincentkam) [reply tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs]: I agree that the fact that the index changed from 1 to 2 indicates that there is a bug.
I restored the variable `lastChangeStrangeDocument` and stepped through this example with a debugger:
`lastChangeStreamDocument.ResumeToken` and `previousCursor.GetResumeToken()` have different values: I'm unsure, but my guess is that it's because the cursor technically fetched two documents and thus should resume from the third, where as the lastChangeStreamDocument (as written in the example) held the document with `{x:1} ` @rstam  Does this explanation sound feasible to you? 

I'm not sure if there's a way to use the cursor's resume token such that the next document that enumerator returns  is `{x:2}`  because our batchsize is  2. If we can't do this, perhaps the example should call `MoveNext` twice as to completely consume the first batch?

**Review by vincentkam — CHANGES_REQUESTED** [review]

File: tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs
Comment (jyemin) [reply tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs:116]: I don't think so.  But let's check with Esha for the necessary workflow between teams.

File: tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs
Comment (rstam) [reply tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs]: The root problem is that the example code is mixing `IAsyncCursor` and `IEnumerator`.

The enumerator is out of sync with the cursor. The enumerator moves one document at a time and the cursor moves one batch at a time.

I recall having a discussion about the interplay between the enumerator and the cursor when `GetResumeToken` was implemented for the C# driver and we decided that we would not support resuming when using `IEnumerator`, only when using `IAsyncCursor`.

So the example needs to be changed to not use `IEnumerator` at all.

Comment (rstam) [reply tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs]: I mentioned when this example was first written that it was not a good example. But we were given some sample code and told to translate verbatim to C# and that's what resulted.

The smallest possible change to this example to get it to work with `GetResumeToken` is to eliminate all use of the enumerator and work directly with the cursor.

It could look like this:

    var documents = new[]
    {
        new BsonDocument("x", 1),
        new BsonDocument("x", 2)
    };

    IChangeStreamCursor<ChangeStreamDocument<BsonDocument>> previousCursor;
    {
        new Thread(() =>
        {
            Thread.Sleep(TimeSpan.FromMilliseconds(100));
            inventory.InsertMany(documents);
        })
        .Start();

        previousCursor = inventory.Watch(new ChangeStreamOptions { BatchSize = 1 });
        while (previousCursor.MoveNext() && previousCursor.Current.Count() == 0) { } // keep calling MoveNext until we've read the first batch
    }

    {
        // Start Changestream Example 3
        var resumeToken = previousCursor.GetResumeToken();
        var options = new ChangeStreamOptions { ResumeAfter = resumeToken };
        var cursor = inventory.Watch(options);
        cursor.MoveNext();
        var next = cursor.Current.First();
        cursor.Dispose();
        // End Changestream Example 3

        next.FullDocument.Should().Be(documents[1]);
    }
}

**Review by rstam — CHANGES_REQUESTED** [review]

File: tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs
Comment (MikalaiMazurenka) [reply tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs]: Done

**Comment by rstam** [comment]
Do the other three examples need to be changed to use `cursor` instead of `enumerator`?

@jyemin ?

**Comment by jyemin** [comment]
What's the advantage to that?  Consistency?

**Comment by rstam** [comment]
Yes, consistency.

And we don't think anyone should be consuming a tailable cursor with an enumerator.

They can if they want I guess, but then they can never change their code to be resumable.

**Review by rstam — CHANGES_REQUESTED** [review]
Waiting to approve until we know if the other three examples need to be changed to use `cursor` instead of `enumerator`.

**Comment by jyemin** [comment]
SGTM

**Comment by MikalaiMazurenka** [comment]
Ready for review

**Review by DmitryLukyanov — APPROVED** [review]

File: tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs
Comment (vincentkam) [inline tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs]: Hm, do you think the comment from example 3 might be helpful here too?  😃 
`// keep calling MoveNext until we've read the first batch`

File: tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs
Comment (vincentkam) [inline tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs]: Hm, do you think the comment from example 3 might be helpful here too?  😃 
`// keep calling MoveNext until we've read the first batch`

File: tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs
Comment (vincentkam) [reply tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs:116]: Spoke with Esha: in theory the docs should be automatically updated once this pull request is merged. 
@MikalaiMazurenka Let's make sure that the updates happen when you do merge this PR: if they don't happen automatically, let's file a docs ticket.

File: tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs
Comment (vincentkam) [inline tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs:164]: This example appears to be missing from https://docs.mongodb.com/manual/changeStreams/#resume-a-change-stream, so we may need a docs ticket if this change doesn't automatically get picked up when we merge this PR.

**Review by vincentkam — CHANGES_REQUESTED** [review]
Minor suggestions made regarding a comment: since these snippets are going to be part of the docs, I think the extra clarifying comment couldn't hurt. What do you think Mikalai?

File: tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs
Comment (MikalaiMazurenka) [reply tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs:164]: Yep, I will keep track of documentation examples after the merge.

File: tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs
Comment (MikalaiMazurenka) [reply tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs]: Done

File: tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs
Comment (MikalaiMazurenka) [reply tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs]: Done

**Comment by MikalaiMazurenka** [comment]
Ready for review

**Review by vincentkam — APPROVED** [review]
LGTM! 😸  (It might be good to wait for Robert's LGTM even though there are already two LGTMs because he was involved with rewriting the examples) :)

File: tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs
Comment (rstam) [inline tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs:49]: These examples are fundamentally flawed.

What application would write code to read ONLY the first change document from a change stream.

What the example *should* be is code showing how to loop while processing multiple change documents.

But... I guess this is a correct implementation of a fundamentally flawed example.

File: tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs
Comment (rstam) [inline tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs:120]: This only works because in the setup to the test we ensured that a change document would be pending.

In general, MoveNext might return false and the call to First in the next statement could throw an exception.

Again... this is because these examples are fundamentally flawed.

**Review by rstam — CHANGES_REQUESTED** [review]
I don't know what to say. Mikalai's changes are fine, but the requested samples are fundamentally flawed.

@jyemin what do you think we should do?

File: tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs
Comment (jyemin) [reply tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs:49]: Robert, I disagree with the severity of your assessment.  Looking at the docs page at https://docs.mongodb.com/manual/changeStreams/ I can see why it was done this way.  There are multiple examples showing different aspects of change streams, and it was probably seen as too distracting to show a loop for each example.

However, the .NET driver is an anomaly in that it exposes a batch-oriented cursor.  In the other driver examples you can imagine, if you know the language, that you would just call next() in a loop using the iteration idiom of the language.  But in the .NET driver it's less obvious what's going on.  The call to `cursor.Current.First()` looks more awkward than the corresponding call to `cursor.next()` in the Java examples.

Given the current constraints, would it make more sense to remove the while loop as well, and just do:

     cursor.MoveNext();
     cursor.Current.First();
     cursor.Dispose();

?  

I think this is also worth a DOCSP ticket expressing your concerns on the overall structure of the examples.  I'm sure they'd appreciate our help in making it better.

Comment (rstam) [reply tests/MongoDB.Driver.Examples/ChangeStreamExamples.cs:49]: That code assumes there is already a change document pending in the change stream.

Otherwise First will throw an exception on an empty batch.

My assertion that the examples are flawed is based on my opinion that no application should *ever* write code that looks like this. By providing this code as an example we are telling our users that this is the way change streams should be processed.

I'll re-review quickly and LGTM since this does in fact correctly implement what we were asked for.

**Review by rstam — APPROVED** [review]

**Pull request closed without merging**
